File: tcp-ip/ch02/tcp_client.cc
//
// Created by Tiancheng on 2024/8/16.
//

#include <arpa/inet.h>
#include <cstdio>
#include <cstdlib>
#include <sys/socket.h>
#include <unistd.h>

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Usage: %s <serverAddr> <serverPort>\n", argv[0]);
        exit(1);
    }

    //* 调用 socket 函数，创建 socket 套接字
    //* 如果继续调用 bind, listen 函数，将成为服务器套接字
    //* 如果继续调用 connect 函数，将成为客户端套接字
    int clientSocketFd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (clientSocketFd == -1) {
        printf("Error created socket\n");
    }

    sockaddr_in serverAddr{};
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = inet_addr(argv[1]);
    serverAddr.sin_port = htons(atoi(argv[2]));

    //* 客户端调用 connect 函数，向服务器发送连接请求
    if (connect(clientSocketFd, (sockaddr *)&serverAddr, sizeof(serverAddr)) ==
        -1) {
        printf("Error connected to server\n");
    }

    //* 客户端调用 read 函数，每次读出一个字符
    char buf[32]; // buffer
    int i = 0, totalBytes = 0;
    while (true) {
        //! 从 clientSocketFd 中读出 1 个字节, 写入 buf
        int readBytes = read(clientSocketFd, &buf[i++], 1);
        if (readBytes <= 0) {
            break;
        }
        totalBytes += readBytes;
    }
    printf("Echo from server: %s\n", buf);
    printf("Message length: %d\n", totalBytes);
    close(clientSocketFd);
    return 0;
}

File: tcp-ip/ch02/go/client/tcp_client.go
package main

import (
	"fmt"
	"io"
	"net"
	"os"
	"strings"
)

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <serverAddr>:<serverPort>\n", os.Args[0])
		os.Exit(1)
	}
	serverAddr := os.Args[1]

	//* 客户端调用 net.Dial 函数，向服务器发送连接请求
	conn, err := net.Dial("tcp", serverAddr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[ERROR] %s\n", err.Error())
		os.Exit(1)
	}

	//* 客户端调用 conn.Read 方法，每次读出一个字符
	buf := make([]byte, 1) // 缓冲区大小为 1
	var msgBuilder strings.Builder
	totalBytes := 0
	for {
		readBytes, err := conn.Read(buf)
		totalBytes += readBytes
		if err == io.EOF {
			break
		}
		msgBuilder.WriteByte(buf[0])
	}
	fmt.Printf("Echo from server: %v\n", msgBuilder.String())
	fmt.Printf("Message length: %v\n", totalBytes)
}
